### PR TITLE
Support matchers for Labels API

### PR DESCRIPTION
**Pull request by yeya24**
This pr is based on #7496 and it adds label names matchers support using the same way.

### CODE REVIEW

File: web/api/v1/api.go [hunk 1]
```diff
@@ -492,14 +492,42 @@ func (api *API) labelNames(r *http.Request) apiFuncResult {
 	if err != nil {
 		return apiFuncResult{nil, &apiError{errorBadData, errors.Wrap(err, "invalid parameter 'end'")}, nil, nil}
 	}
+	var matcherSets [][]*labels.Matcher
+	for _, s := range r.Form["match[]"] {
```
Comment (brian-brazil) [inline web/api/v1/api.go:496]: We now need to complete the discussion that was cut short in #8288, as this should have the same handling of empty selectors as all the other places.

File: web/api/v1/api.go [hunk 2]
```diff
@@ -492,14 +492,42 @@ func (api *API) labelNames(r *http.Request) apiFuncResult {
...
+			Func:  "series", // There is no series function, this token is used for lookups that don't need samples.
+		}
+
+		// Get all series which match matchers.
+		var sets []storage.SeriesSet
+		for _, mset := range matcherSets {
+			s := q.Select(false, hints, mset...)
+			sets = append(sets, s)
+		}
+		// TODO(yeya24): match these at TSDB level.
```
Comment (brian-brazil) [inline web/api/v1/api.go:525]: Why do you want to do that? I'm not seeing how that would be notably any more efficient, without duplicating the entirety of the rather intricate postings matching logic which I don't think would be wise.

File: web/api/v1/api.go [hunk 3]
```diff
@@ -553,6 +609,53 @@ func (api *API) labelValues(r *http.Request) (result apiFuncResult) {
...
+	labelValues := make([]string, 0, len(labelValuesSet))
+	for key := range labelValuesSet {
+		labelValues = append(labelValues, key)
+	}
+	sort.Strings(labelValues)
+	return labelValues, warnings, nil
+}
+
+// labelNamesByMatchers uses matchers to filter out matching series, then label names are extracted.
+func labelNamesByMatchers(sets []storage.SeriesSet) ([]string, storage.Warnings, error) {
```
Comment (brian-brazil) [inline web/api/v1/api.go:636]: This is only used once, and is not close to where it is used making understanding the code hard. Particularly as the function name is misleading as this code has nothing to do with matchers.

File: web/api/v1/api.go [hunk 4]
```diff
@@ -553,6 +609,53 @@ func (api *API) labelValues(r *http.Request) (result apiFuncResult) {
...
+	set := storage.NewMergeSeriesSet(sets, storage.ChainedSeriesMerge)
```
Comment (brian-brazil) [inline web/api/v1/api.go:637]: Would we be faster iterating over each set separately?

File: web/api/v1/api.go [hunk 5]
```diff
@@ -553,6 +609,53 @@ func (api *API) labelValues(r *http.Request) (result apiFuncResult) {
 	return apiFuncResult{vals, nil, warnings, closer}
 }
 
+// LabelValuesByMatchers uses matchers to filter out matching series, then label values are extracted.
+func labelValuesByMatchers(sets []storage.SeriesSet, name string) ([]string, storage.Warnings, error) {
+	set := storage.NewMergeSeriesSet(sets, storage.ChainedSeriesMerge)
+	labelValuesSet := make(map[string]struct{})
+	for set.Next() {
+		series := set.At()
+		labelValue := series.Labels().Get(name)
```
Comment (brian-brazil) [inline web/api/v1/api.go:618]: You need to filter out empty here somewhere.

File: web/api/v1/api.go (hunk 2 above)
Comment (yeya24) [reply web/api/v1/api.go:525]: Yeah, I agree. I will remove this comment.

**Comment by yeya24** [comment]
Hi, @brian-brazil  I updated the pr. Please take a look

File: docs/querying/api.md [hunk 6]
```diff
@@ -271,6 +271,8 @@ URL query parameters:
 
 - `start=<rfc3339 | unix_timestamp>`: Start timestamp. Optional.
 - `end=<rfc3339 | unix_timestamp>`: End timestamp. Optional.
+- `match[]=<series_selector>`: Repeated series selector argument that selects the
```
Comment (brian-brazil) [inline docs/querying/api.md:274]: Optional

File: web/api/v1/api.go (hunk 4 above)
Comment (brian-brazil) [reply web/api/v1/api.go:633]: This wasn't addressed.

File: web/api/v1/api.go [hunk 7]
```diff
@@ -553,6 +601,57 @@ func (api *API) labelValues(r *http.Request) (result apiFuncResult) {
...
+	labelValues := make([]string, 0, len(labelValuesSet))
+	for key := range labelValuesSet {
+		labelValues = append(labelValues, key)
+	}
+	sort.Strings(labelValues)
+	return labelValues, warnings, nil
+}
+
+// getLabelNamesFromSeriesSet extracts label names from matching series.
+func getLabelNamesFromSeriesSet(sets []storage.SeriesSet) ([]string, storage.Warnings, error) {
```
Comment (brian-brazil) [inline web/api/v1/api.go:632]: This should be inlined, I shouldn't have to jump around the file to find how this all works.

File: web/api/v1/api.go [hunk 8]
```diff
@@ -1631,6 +1713,31 @@ func parseDuration(s string) (time.Duration, error) {
...
+			return nil, err
+		}
+		matcherSets = append(matcherSets, matchers)
+	}
+
+	for _, ms := range matcherSets {
+		var nonEmpty bool
+		for _, lm := range ms {
+			if lm != nil && !lm.Matches("") {
+				nonEmpty = true
```
Comment (brian-brazil) [inline web/api/v1/api.go:1730]: You could also continue the outer for

File: docs/querying/api.md (hunk 6 above)
Comment (yeya24) [reply docs/querying/api.md:274]: Updaed.

File: web/api/v1/api.go (hunk 4 above)
Comment (yeya24) [reply web/api/v1/api.go]: I updated it for the `LabelNames` part, do I need to do the same thing for `LabelValues`?

File: web/api/v1/api.go (hunk 7 above)
Comment (yeya24) [reply web/api/v1/api.go]: Done.

File: web/api/v1/api.go (hunk 8 above)
Comment (yeya24) [reply web/api/v1/api.go]: Done.

File: web/api/v1/api.go (hunk 4 above)
Comment (brian-brazil) [reply web/api/v1/api.go]: It's more a question of which is more efficient? The chain will use more RAM anyway, and we should be consistent.

Comment (yeya24) [reply web/api/v1/api.go]: Yes, you are right, using `MergeSeriesSet` uses more RAM.

Screenshot from 2020-12-18 13-40-01

Comment (yeya24) [reply web/api/v1/api.go]: Updated, PTAL @brian-brazil

**Comment by yeya24** [comment]
Can I get another round of review of this pr?

**Comment by brian-brazil** [comment]
I'll take a look tomorrow.

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**